gh-128965: pickle `load_build` function checks if `state` is None, not False

Legoclones opened the pull request:
Inside of the `load_build()` function for pickle's `BUILD` opcode, the C accelerator at one point checks if `state` is `Py_None`, while the Python version only checks `if state`. 

https://github.com/python/cpython/blob/34ded1a1a10204635cad27830fcbee2f8547e8ed/Modules/_pickle.c#L6638

https://github.com/python/cpython/blob/34ded1a1a10204635cad27830fcbee2f8547e8ed/Lib/pickle.py#L1765

This means if `state` is something like an empty dictionary or tuple, the code block under the `if` statement WILL be run in `_pickle.c`, but NOT in `pickle.py`. 

As an example, the bytestream `b']]b.'` has the following disassembly:
```python
    0: ]    EMPTY_LIST
    1: ]    EMPTY_LIST
    2: b    BUILD
    3: .    STOP
highest protocol among opcodes = 1
```

This will do nothing in `pickle.py` but error out in `_pickle.c` with the message `state is not a dictionary`. The easy solution is to change `if state` to `if state != None`, and it shouldn't break any existing functionality. I've attached a pull request. 

<!-- gh-issue-number: gh-128965 -->
* Issue: gh-128965
<!-- /gh-issue-number -->

Legoclones committed:
`load_build` checks if `state` is None, not False

picnixz commented on Lib/pickle.py:
```diff
@@ -1837,7 +1837,7 @@ def load_build(self):
         if isinstance(state, tuple) and len(state) == 2:
             state, slotstate = state
-        if state:
+        if state != None:
```
```suggestion
        if state is not None:
```

picnixz reviewed (commented):
Can we have tests that check this code path? Namely check that non-dict objects raise a TypeError and that falsey dicts do not raise exceptions? In addition, we need a NEWS entry for the bugfix, e.g:

```rst
Fix :meth:`!pickle._Pickler.load_build` for non-dictionary states.
```

(My NEWS entry is definitely poorly worded but I don't have a better suggestion for now. Maybe you can come up with a better formulation).

Legoclones committed:
Update Lib/pickle.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

Legoclones committed:
Add test_load_build_falsey_state

Added 6 tests where the `state` is falsey but not None, and the `inst` value is invalid - these should throw an AttributeError because the `__dict__` attribute doesn't exist for `inst`. In `pickle.py`, if `state` is falsey but not None code is never run to check that `inst` is a valid object (but it does happen in the C accelerator).

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Legoclones committed:
Change pickle tests to use Python loads

Legoclones committed:
Fix backticks

Legoclones committed:
Fixed tests

Legoclones committed:
Merge branch 'main' into load-build-state
